### PR TITLE
Refactor: Implement Singleton ModelManager to fix VRAM memory leak

### LINKED ISSUE #520: [Enhancement]: Fix redundant model loading with Singleton ModelManager
### The Problem Context

Currently in backend/Generator/main.py, massive machine learning models and NLP tools are being initialized independently inside the __init__ methods of MCQGenerator, ShortQGenerator, and ParaphraseGenerator.

Because server.py creates instances of these classes globally, it forces Python to load t5-large (which is ~3GB) and en_core_web_sm into memory three separate times at startup.

- This consumes roughly ~9GB of RAM/VRAM just to hold redundant models.
- It makes server startup painfully slow.
- It guarantees a CUDA "Out of Memory" (OOM) crash for anyone trying to run or host the backend on a smaller GPU (e.g., 8GB VRAM).

### Proposed Solution

I propose implementing a Singleton ModelManager to act as a single source of truth for the heavy NLP models.

**Implementation Steps:**

- Create a ModelManager class that uses the __new__ method to ensure it is only ever instantiated once.
- Load t5-large, spacy, Sense2Vec, and FreqDist exactly once inside this manager.
- Refactor the __init__ methods of the generator classes so they request the existing ModelManager instance and simply hold lightweight pointers to the shared models in memory.

Here is the exact implementation I have tested:

Image

(The __init__ methods of the generator classes are then refactored to simply call manager = ModelManager() and assign self.model = manager.qg_model).

### Impact & Testing

This architectural change is purely internal plumbing and does not change the external API behavior or endpoint logic.

- Memory footprint: Drops from ~9GB down to ~3GB.
- Server startup: Drastically faster since models are read from disk only once.
- Status: I have already built and tested this locally. All endpoints return 200 OK and pass the assertions in test_server.py with identical output logic.

### Code of Conduct

- [x] I have joined the [Discord server](https://discord.gg/hjUhu33uAn) and will post updates there
- [x] I have searched existing issues to avoid duplicates

### PR DESCRIPTION
**Pull request by Siddhazntx**
###  Addressed Issues:

Fixes #520 

###  Screenshots/Recordings:

**Server Startup Output** :

```
> [nltk_data]    |
> [nltk_data]  Done downloading collection popular
> Starting Flask App...
> Initializing Shared ModelManager... Loading massive models into memory ONCE.
> You are using the default legacy behaviour of the <class 'transformers.models.t5.tokenization_t5.T5Tokenizer'>.
```

###  Additional Notes:

**Summary of Changes:**

- Memory Optimization: Implemented a ModelManager Singleton in main.py to prevent server.py from loading the 3GB t5-large model and spacy tools multiple times into RAM/VRAM.
- Refactoring: Updated MCQGenerator, ShortQGenerator, and ParaphraseGenerator to utilize lightweight pointers to the shared models via the ModelManager. This drops the overall memory footprint from ~9GB down to ~3GB and significantly speeds up server boot times.
- Test Fix: Updated the stale /get_answer endpoint string in test_server.py to the correct /get_shortq_answer route. All backend tests now pass locally without any Out-Of-Memory (OOM) crashes.

This is strictly an internal architectural improvement; the external API behavior remains completely unchanged.

## Checklist

- [x] My PR addresses a single issue, fixes a single bug or makes a single improvement.
- [x] My code follows the project's code style and conventions
- [ ] If applicable, I have made corresponding changes or additions to the documentation
- [ ] If applicable, I have made corresponding changes or additions to tests
- [x] My changes generate no new warnings or errors
- [x] I have joined the [Discord server](https://discord.gg/hjUhu33uAn) and I will share a link to this PR with the project maintainers there
- [x] I have read the [Contribution Guidelines](../CONTRIBUTING.md)
- [x] Once I submit my PR, CodeRabbit AI will automatically review it and I will address CodeRabbit's comments.

## AI Usage Disclosure

Check one of the checkboxes below:

- [x] This PR does not contain AI-generated code at all.
- [ ] This PR contains AI-generated code. I have tested the code locally and I am responsible for it.

## Summary by CodeRabbit

* **Refactor**
  * Share heavy ML resources across question generators to reduce startup cost and improve inference consistency
  * Improved device handling and disabled unnecessary gradient tracking for more reliable, faster predictions

* **Chores**
  * Renamed test API endpoint for short-question answers to better reflect its purpose

### CODE REVIEW

**Comment by Siddhazntx** [comment]
@coderabbitai , I have addressed both of these findings in the latest commit:

1. Thread-Safety: Added a class-level threading.Lock() and implemented double-checked locking in both __new__ and __init__ within the ModelManager class. This ensures heavy model initialization is strictly single-threaded and prevents race conditions.
2. Test Naming: Renamed test_get_answer to test_get_shortq_answer in test_server.py and updated the corresponding print statements to accurately reflect the endpoint being tested.

Thanks for catching these!

**Comment by SxBxcoder** [comment]
This PR attempts to address the redundant model initialization issue, expanding on the thread-safety and double-checked locking mechanisms I established in #473. However, after a technical audit of the architecture, there are a few major blockers we need to address before this can be safely merged:

* **The Singleton Anti-Pattern:** Overriding `__new__` and `__init__` to force a Singleton class is highly brittle in Python, especially within a Flask environment that may scale to multi-worker deployments (like Gunicorn). A module-level cache or a strictly static class is a much safer, more Pythonic way to handle global state without fighting the object instantiation cycle.
* **Incomplete PyTorch Optimization:** I see `model.eval()` and `@torch.no_grad()` were applied to the `AnswerPredictor`. However, these critical VRAM-saving optimizations are missing from the actual generation calls in `MCQGenerator`, `ShortQGenerator`, etc. If the goal is to fix VRAM memory leaks, the auto-regressive `generate()` calls must also execute within a `no_grad` context, as they consume the most memory.
* **Architectural Bloat (`main.py`):** Injecting a massive `ModelManager` class directly into `main.py` violates the Single Responsibility Principle and severely bloats the file. If we are moving toward centralized management, this logic needs to be extracted into its own dedicated module.

I recommend we pause on merging this specific Singleton implementation until we can design a cleaner, non-Singleton module level cache that properly wraps *all* inference calls in `torch.no_grad()` to truly solve the VRAM spikes.
